packages/financial-templates-lib/test/price-feed/ExpressionPriceFeed.js

```diff
@@ -0,0 +1,163 @@
+    assert.equal(expressionPriceFeed.getLastUpdateTime(), null);
+  });
+
+  it("undefined inputs", async function() {
+    const priceFeeds = {
+      //                currentPrice      historicalPrice    lastUpdatedTime
+      ETHUSD: new PriceFeedMock(toBN(toWei("1")), toBN(toWei("17")), 100),
+      BTCUSD: new PriceFeedMock(undefined, undefined, undefined)
+    };
+
+    const expressionPriceFeed = new ExpressionPriceFeed(priceFeeds, "ETHUSD * BTCUSD * COMPUSD");
+
+    // Should return null since there was an undefined price output.
+    assert.equal(expressionPriceFeed.getCurrentPrice(), null);
+
+    // Should throw since there was an undefined price output.
+    const arbitraryHistoricalTimestamp = 1000;
+    assert.isTrue(await expressionPriceFeed.getHistoricalPrice(arbitraryHistoricalTimestamp).catch(() => true));
+
+    // Should return null since there was an undefined output.
+    assert.equal(expressionPriceFeed.getLastUpdateTime(), null);
+  });
+
+  it("Validates feeds decimals correctly", async function() {
+    const validDecimalsPriceFeeds = {
+      //                        currentPrice      historicalPrice    lastUpdatedTime   decimals
+      ETHUSD: new PriceFeedMock(toBN(toWei("1")), toBN(toWei("25")), 100, 18),
+      BTCUSD: new PriceFeedMock(toBN(toWei("2")), toBN(toWei("57")), 50000, 18, 4000),
+      COMPUSD: new PriceFeedMock(toBN(toWei("9")), toBN(toWei("10")), 25, 18, 50000)
+    };
+
+    const validExpressionPriceFeed = new ExpressionPriceFeed(validDecimalsPriceFeeds, "ETHUSD * BTCUSD * COMPUSD");
+    assert.equal(validExpressionPriceFeed.getPriceFeedDecimals(), 18);
+
+    // Create three feeds, one with a diffrent number of decimals. Medianizer should reject this when checking the decimals.
+    const inValidDecimalsPriceFeeds = {
+      //                        currentPrice      historicalPrice    lastUpdatedTime   decimals
+      ETHUSD: new PriceFeedMock(toBN(toWei("1")), toBN(toWei("25")), 100, 18),
+      BTCUSD: new PriceFeedMock(toBN(toWei("2")), toBN(toWei("57")), 50000, 18),
+      COMPUSD: new PriceFeedMock(toBN(toWei("9")), toBN(toWei("10")), 25, 17)
+    };
+
+    const invalidExpressionPriceFeed = new ExpressionPriceFeed(inValidDecimalsPriceFeeds, "ETHUSD * BTCUSD * COMPUSD");
+    assert.throws(() => invalidExpressionPriceFeed.getPriceFeedDecimals());
+  });
+
+  describe("Expression parsing with escaped characters", async function() {
+    it("Escapes characters correctly", async function() {
+      assert.equal(escapeSpecialCharacters("USD-[bwBTC/ETH SLP]"), "USD\\-\\[bwBTC\\/ETH\\ SLP\\]");
+      assert.equal(escapeSpecialCharacters("-/][a--]rr]"), "\\-\\/\\]\\[a\\-\\-\\]rr\\]");
+    });
+
+    it("Processes escaped expressions correctly", async function() {
+      const escapedSymbol = "\\[1\\/\\-2\\]\\ \\-\\ 17";
+      assert.equal(math.evaluate(`${escapedSymbol} - 2 * ${escapedSymbol}`, { [escapedSymbol]: 5 }).toString(), "-5");
+
+      const unescapedSymbol = "USD-[bwBTC/ETH SLP]";
```

Comment:
would have been good for the tests above to use this identifier in one of the expressions

Reply:
Done! Added it to one of the earlier tests.